stan/math/prim/fun/interp1.hpp

```diff
@@ -1,45 +1,48 @@
 #ifndef STAN_MATH_PRIM_FUN_INTERP1_HPP
 #define STAN_MATH_PRIM_FUN_INTERP1_HPP
 
-#include <stan/math/prim/meta.hpp>
+//#include <stan/math/prim/fun/sort_asc.hpp>
+
 using namespace std;
 namespace stan {
     namespace math {
-    /**
-     * Return the one-dimensional interpolation of the second argument of points corresponding to the first argument.
-     * @tparam T_true type of the true argument
-     * @tparam T_false type of the false argument
-     * @param c Boolean condition value.
-     * @param y_true Value to return if condition is true.
-     * @param y_false Value to return if condition is false.
-     */
-    inline Eigen::VectorXd
-    interp1(const Eigen::VectorXd& xData, const Eigen::VectorXd& yData, const Eigen::VectorXd& xVector)
-    {
-
-        Eigen::VectorXd yVals;
-        int size_x = xData.size();
-        int size_xData = xData.size();
+        template <typename T>
+        inline std::vector<double> interp1(const std::vector<T>& xDataPre, const std::vector<T>& yDataPre, const std::vector<T>& xTest)
+        {
+            using stan::math::index_type_t;
+            int size_xData = xDataPre.size();
+            std::vector<double> dydx(size_xData);
+            int size_x = xTest.size();
+            std::vector<double> yVals(size_x);
 
-        for(int j=0; j < size_x; j++) {
-            double x = xVector[j];
-            int i = 0;
-            if (x >= xData[size_xData - 2])                                           // special case: beyond right end
-            {
-                i = size_xData - 2;
-            } else {
-                while (x > xData[i + 1])
-                    i++;                                   // find left end of interval for interpolation
+            // sort if xData is not monotone
+            std::vector< pair<int, int> > vec;
+            for(int i=0; i<size_xData;i++){
+                vec.push_back(make_pair(xDataPre[i], yDataPre[i]));
+            }
+            std::sort(vec.begin(), vec.end());
+            std::vector<double > xData(size_xData);
+            std::vector<double > yData(size_xData);
+            for (int i=0; i<size_xData; i++){
+                xData[i] = vec[i].first;
+                yData[i] = vec[i].second;
+            }
+            for (int i = 0; i < size_xData; i++){
```

Comment:
There's an off by one error here. So since xData is size `size_xData` (I find it convenient to use single capital letters like `N` or `M` for these sorts of things), this loop should be:

```
for (int i = 0; i < size_xData - 1; i++)
```

to avoid overflow.